prototype_source/pt2e_quantizer.rst

```diff
@@ -302,6 +302,114 @@ functions that are used in the example:
+
+One caveat is that if inputs of the pattern has multiple users, we don't have a good way to identify which user node we want to annotate except for checking the aten op target.
+
+Another caveat is that we need to make sure we have an exhaustive list of examples (e.g. 2D, 3D, 4D inputs, real v.s. symbolic inputs, training=True v.s. training=False etc.) for the pattern to make sure cover different possible ``aten`` IR outcomes captured from the ``torch`` IR pattern.
+
+3. Using ``SubgraphMatcherWithNameNodeMap``
+----------------------------------------------
+We also introduced a different SubgraphMatcher util called ``SubgraphMatcherWithNameNodeMap`` to make it easier to query the nodes that people want to annotate.
+
+Example::
+  
+  def conv_relu_pattern(x, weight, bias):
+      conv = torch.nn.functional.conv2d(x, weight, bias)
+      relu = torch.nn.functional.relu(conv)
+      # returns an additional dict that includes a map from name to node that we want to annotate
+      return relu, {"conv": conv, "relu": relu, "x": x, "weight": weight, "bias": bias}
+
+  matcher = SubgraphMatcherWithNameNodeMap(conv_relu_pattern)
+  matches = matcher.match(model)
+  for match in matches:
+      # find input and output of the pattern
+      # annotate the nodes
+      name_node_map = match.name_node_map
+      name_node_map["conv"].meta["quantization_annotation"] = ...
+      name_node_map["relu"].meta["quantization_annotation"] = ...
+
+This should be easier to use than the original ``SubgraphMatcher`` around finding the node we want to annotate or use. However, it may not work if some of the operators are captured as multiple ops. For example, if ``torch.nn.functional.conv2d`` is captured as multiple ops, then the inputs ``x``, ``weight``, ``bias`` may not be the direct inputs to the final operator marked by ``conv``, in this case we'll need to revert to the same way of access mentioned in the previous example.
+
+Example::
+
+  for match in matches:
+      # find input and output of the pattern
+      # annotate the nodes
+      name_node_map = match.name_node_map
+      input_x = name_node_map["x"]
+      # since input_x may not be the input of name_node_map["conv"]
+      # we'll get the user nodes of this op and annotate that instead
+      input_x.users[0].meta["quantization_annotation"] = ...
+      ...
+
+
+Similar to ``SubgraphMatcher``, we also need to make sure to capture the PyTorch pattern with different example inputs to cover all possible ``aten`` IR variants.
+
+4. General Recommentation
+----------------------------------------------
+Given the above UX and constraints/caveats, we recommend people to start with option 3. and make sure the first operator in the pattern are functional ops that map to one aten op (e.g. F.conv2d, F.linear), and provide enough variants of example inputs. (We may provide some (pattern, list of example_inputs) so people can just use them directly in the future)
```

Comment:
I would really really stay away from "recommending". People implementing Quantizer should make their choice. When you "recommend" it feels like you are implicitly supporting something, and when that doesnt work, "people" will come to you complaining.

We should just highlight different ways of graph matching, that is it. And this is not specific to quantization, so ideally this should have its own user manual somewhere.

Reply:
I do plan to provide some utils actually, since now we separated patterns matching and annotation, we could maintain a centralized place for patterns

still don't quite following the "recommendation" point, but I feel at this point looks like we can just recommend one API

Reply:
Actually now that I think of it, we should actually do a pattern cache of some sort that can be used not just for quantization but also for delegation.

Reply:
yes, Greg also mentioned this before, we can probably capture some of these patterns beforehand and users can use them for matching directly. maybe these could be used for delegation as well if you want to match the pattern and then expand the pattern to cover q/dq ops etc.